### PR TITLE
[flutter_tools] Clean up `boolArgDeprecated` and `stringArgDeprecated`

### LINKED ISSUE #105263: [flutter_tools] boolArg() should be non-nullable, and throw if the flag doesn't exist
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1526

Discussion: https://github.com/flutter/flutter/pull/105242#discussion_r888395925

### PR DESCRIPTION
**Pull request by bartekpacia**
This PR standarizes the methods used to parse command-line flags and options. 

Fixes #105263 as discussed in https://github.com/flutter/flutter/issues/105263#issuecomment-1459093624.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by bartekpacia** [comment]
The 3 test failures (`Mac tool_tests_commands`, `Windows tool_tests_commands`, `tool_tests-commands-linux`) are coming from the single group in `test/commands.shard/hermetic/run_test.dart` – `usageValues`. I think there's a bug in that test that became visible just now that  `boolArg` throws if `argResults` is null instead of falling back to `null`.

It crashes in these lines: [676](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/commands.shard/hermetic/run_test.dart#L676), [699](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/commands.shard/hermetic/run_test.dart#L699), and a couple of the same ones in `usageValues` group.

Crashes in these lines come from [here in `RunCommand.usageValues`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/run.dart#L487).

I fiddled with it for half an hour or so but I didn't find a nice way to fix it.

@christopherfujino @jmagman

**Comment by bartekpacia** [comment]
Oh, I think I found [a fix](https://github.com/flutter/flutter/pull/122184/commits/06192ea0cec84a78b421b994292d001a6276aaa4) :)

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 1]
```diff
@@ -1712,27 +1712,17 @@ abstract class FlutterCommand extends Command<void> {
...
-    if (!argParser.options.containsKey(name)) {
-      return null;
-    }
-    return argResults![name] as bool;
-  }
+  /// Gets the parsed command-line flag named [name] as a `bool`.
+  ///
+  /// If no flag named [name] was parsed (either from the command-line or from
+  /// defaults), an [ArgumentError] is thrown.
+  bool boolArg(String name) => argResults![name] as bool? ?? false;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1719]:
```suggestion
  bool boolArg(String name) => argResults![name] as bool;
```

File: packages/flutter_tools/lib/src/commands/drive.dart [hunk 2]
```diff
@@ -73,7 +73,6 @@ class DriveCommand extends RunCommandBase {
     addMultidexOption();
     argParser
       ..addFlag('keep-app-running',
-        defaultsTo: null,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/drive.dart:76]: hmm, interesting. So we DID hit the fallback in `boolArgDeprecated`. In any case, having the flag default to `false` is better, so good cleanup.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 3]
```diff
@@ -1712,27 +1712,17 @@ abstract class FlutterCommand extends Command<void> {
...
-  bool? boolArg(String name) {
-    if (!argParser.options.containsKey(name)) {
-      return null;
-    }
-    return argResults![name] as bool;
-  }
+  /// Gets the parsed command-line flag named [name] as a `bool`.
+  ///
+  /// If no flag named [name] was parsed (either from the command-line or from
+  /// defaults), an [ArgumentError] is thrown.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1718]:
```suggestion
  /// If no flag named [name] was added to the [ArgParser], an [ArgumentError]
  /// will be thrown.
```

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 4]
```diff
@@ -1712,27 +1712,17 @@ abstract class FlutterCommand extends Command<void> {
...
-
-  String? stringArg(String name) {
-    if (!argParser.options.containsKey(name)) {
-      return null;
-    }
-    return argResults![name] as String?;
-  }
+  ///
+  /// If no option named [name] was parsed (either from the command-line or from
+  /// defaults), an [ArgumentError] is thrown.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1724]:
```suggestion
  /// If no option named [name] was added to the [ArgParser], an [ArgumentError]
  /// will be thrown.
```

File: packages/flutter_tools/test/commands.shard/hermetic/run_test.dart [hunk 5]
```diff
@@ -1173,6 +1173,9 @@ class TestRunCommandForUsageValues extends RunCommand {
   // ignore: overridden_fields
   List<Device>? devices;
 
+  @override
+  bool get enableImpeller => false;
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:1177]: Why was this added?

Comment (bartekpacia) [reply packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:1177]: I described this problem in more details [here](https://github.com/flutter/flutter/pull/122184#issuecomment-1460508677). Here's a [test failure](https://github.com/flutter/flutter/runs/11854897675).

I think this was a mistake in all tests in [the usageValues group](https://github.com/flutter/flutter/blob/2b6a79cca5e45ce23cade706bbf729b44b57ef03/packages/flutter_tools/test/commands.shard/hermetic/run_test.dart#L670), but they were passing somewhat by accident because `boolArg` was falling back to `false`:

```dart
bool boolArg(String name) => argResults?[name] as bool? ?? false;
```

Now that we've changed `boolArg` to:

```dart
bool boolArg(String name) => argResults![name] as bool;
```

the tests started throwing, because they (the tests) don't call `Command.run()` so `argResults` are null.

File: packages/flutter_tools/lib/src/commands/analyze.dart [hunk 6]
```diff
@@ -144,12 +144,12 @@ class AnalyzeCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
-    final bool? suggestionFlag = boolArg('suggestions');
-    final bool machineFlag = boolArg('machine') ?? false;
-    if (suggestionFlag != null && suggestionFlag == true) {
+    final bool suggestionFlag = boolArg('suggestions');
+    final bool machineFlag = boolArg('machine');
+    if (suggestionFlag == true) {
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/flutter_tools/lib/src/commands/analyze.dart]: Thanks. BTW, shouldn't you enable a lint rule for redundant comparisons like this one?

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 7]
```diff
@@ -191,23 +191,23 @@ known, it can be explicitly provided to attach via the command-line, e.g.
     if (argResults!['debug-url'] == null) {
       return null;
     }
-    final Uri? uri = Uri.tryParse(stringArgDeprecated('debug-url')!);
+    final Uri? uri = Uri.tryParse(stringArg('debug-url')!);
     if (uri == null) {
-      throwToolExit('Invalid `--debug-url`: ${stringArgDeprecated('debug-url')}');
+      throwToolExit('Invalid `--debug-url`: ${stringArg('debug-url')}');
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/flutter_tools/lib/src/commands/attach.dart:196]: thanks, done :)

File: packages/flutter_tools/lib/src/commands/drive.dart (hunk 2 above)
Comment (bartekpacia) [reply packages/flutter_tools/lib/src/commands/drive.dart:76]: Yeah, this was the only such occurrence in the whole codebase, kinda bizarre.

File: packages/flutter_tools/test/commands.shard/hermetic/run_test.dart (hunk 5 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:1177]: ahh, sorry, i missed reading that comment. my fault.

**Comment by bartekpacia** [comment]
Hmm, looks like the `Mac module_custom_host_app_name_test` [has timed out](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8787159192147060977/+/u/run_module_custom_host_app_name_test/execution_details). Don't know why's that.

File: packages/flutter_tools/test/commands.shard/hermetic/run_test.dart (hunk 5 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/commands.shard/hermetic/run_test.dart]: pushed a patch that invokes the command first to ensure the arg is passed

Comment (bartekpacia) [reply packages/flutter_tools/test/commands.shard/hermetic/run_test.dart]: Thanks, looks cleaner now B)

**Review by christopherfujino — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/commands/analyze.dart [hunk 8]
```diff
@@ -144,12 +144,12 @@ class AnalyzeCommand extends FlutterCommand {
...
-    final bool machineFlag = boolArg('machine') ?? false;
-    if (suggestionFlag != null && suggestionFlag == true) {
+    final bool suggestionFlag = boolArg('suggestions');
+    final bool machineFlag = boolArg('machine');
+    if (suggestionFlag) {
       final String directoryPath;
-      final bool? watchFlag = boolArg('watch');
-      if (watchFlag != null && watchFlag) {
+      final bool watchFlag = boolArg('watch');
+      if (watchFlag) {
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/commands/analyze.dart:152]: Nit: consider removing `watchFlag` and moving boolArg('watch') into the `if`

**Review by andrewkolos — APPROVED** [review]
LGTM with nit

File: packages/flutter_tools/lib/src/commands/analyze.dart (hunk 8 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/analyze.dart:152]: Could also do the same with `suggestionFlag` and `machineFlag` since these are all referenced only once

Comment (bartekpacia) [reply packages/flutter_tools/lib/src/commands/analyze.dart]: Done, thanks!

**Comment by bartekpacia** [comment]
Yup, thanks for heads up, fixed :)

**Comment by bartekpacia** [comment]
> Google testing failed. Please contact a Google developer to investigate.

Oopsie, @christopherfujino?
